examples/model_selection/plot_successive_halving_heatmap.py

```diff
@@ -0,0 +1,117 @@
+"""
+Comparison between grid search and successive halving
+=====================================================
+
+This example compares the parameter search performed by
+:class:`~sklearn.model_selection.HalvingGridSearchCV` and
+:class:`~sklearn.model_selection.GridSearchCV`.
+
+The heatmap shows the mean test score of the parameter combinations for an
+SVC instance. The
+:class:`~sklearn.model_selection.HalvingGridSearchCV`
+also shows the iteration at which the combinations where last used. The
+combinations marked as ``0`` were only evaluated at the first iteration, while
+the ones with ``5`` are the parameter combinations that are considered the
+best ones.
+
+We can see that the :class:`~sklearn.model_selection.HalvingGridSearchCV`
+class is able to find parameter combinations that are just as accurate as
+:class:`~sklearn.model_selection.GridSearchCV`, in much less time.
+"""
+from time import time
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+from sklearn.svm import SVC
+from sklearn import datasets
+from sklearn.model_selection import GridSearchCV
+from sklearn.model_selection import HalvingGridSearchCV
+
+
+print(__doc__)
+
+rng = np.random.RandomState(0)
+X, y = datasets.make_classification(n_samples=1000, random_state=rng)
+
+gammas = [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7]
+Cs = [1, 10, 100, 1e3, 1e4, 1e5]
+param_grid = {'gamma': gammas, 'C': Cs}
+
+clf = SVC(random_state=rng)
+
+# run HalvingGridSearchCV
+tic = time()
+gsh = HalvingGridSearchCV(
+    estimator=clf,
+    param_grid=param_grid,
+    resource='n_samples',
+    max_resources='auto',  # max_resources=n_samples
+    force_exhaust_resources=True,
+    cv=5,
+    ratio=2,
+    random_state=rng)
+gsh.fit(X, y)
+gsh_time = time() - tic
+
+# run GridSearchCV
+tic = time()
+gs = GridSearchCV(
+    estimator=clf,
+    param_grid=param_grid,
+    cv=5)
+gs.fit(X, y)
+gs_time = time() - tic
+
+
+def make_heatmap(ax, gs, show_iter=False, make_cbar=False):
+    """Helper to make a heatmap."""
+    results = pd.DataFrame.from_dict(gs.cv_results_)
+    results['params_str'] = results.params.apply(str)
+    # Take max but there's only one value anyway
+    scores = results.groupby(['param_gamma', 'param_C']).mean_test_score.max()
+    scores_matrix = scores.values.reshape(len(gammas), len(Cs))
```

Comment:
hmm, unstack doesn't work because there are duplicated indices.

Which made me realize that the code above and the comment are incorrect: we want a 2D grid whose values for `mean_test_score` correspond to the highest `iter`. Right now it takes the max of `mean_test_score` without considering `iter`.

My pandas is quite weak, any tips @jnothman?

Reply:
I don't find pandas nice to express these things in.

This is one way to get `iter` and `values` with params as index:
```
df.sort_values('iter').groupby(['param_gamma', 'param_C']).agg('last')
```
this is another (although it also has a params column in its output). Uglier, but more efficient:
```
df.groupby(['param_gamma', 'param_C']).apply(lambda x: x.loc[x.iter == x.iter.max()].iloc[0])
```

with either of these you could do `.mean_test_score.unstack()` or similar.